src/log.js

```diff
@@ -1,153 +1,389 @@
+   * @returns {Log}
+   */
+  static joinAll(ipfs, logs) {
+    return logs.reduce((log, val, i) => {
+      if (!log) return val
+      return LogUtils.join(ipfs, log, val)
+    }, null)
+  }
+
+  /**
+   * Expand the size of the log
+   * @param {IPFS} [ipfs] An IPFS instance
+   * @param {Log} log
+   * @param {Number} length
+   * @param {function(hash, entry, parent, depth)} onProgressCallback
+   * @returns {Promise<Log>}
+   */
+  static expand(ipfs, log, length = -1, onProgressCallback) {
+    // TODO: Find tails (entries that point to an entry that is not in the log)
+    const tails = log.items.slice()[0].next.sort(LogUtils._compare)
+    // Fetch a log starting from each tail entry
+    const getLog = tails.map((f) => LogUtils.fromEntry(ipfs, f, length, onProgressCallback))
+    // Join all logs together to one log
+    const joinAll = (logs) => LogUtils.joinAll(ipfs, logs.concat([log]))
+    // Create all logs and join them
+    return Promise.all(getLog).then(joinAll)
+  }
+
+  /**
+   * Insert an entry to the log
+   * @private
+   * @param {Entry} entry Entry to be inserted
+   * @returns {Entry}
+   */
+  static _insert(ipfs, log, entry) {
+    const hashes = log.items.map((f) => f.hash)
+    // If entry is already in the log, don't insert
+    if (hashes.includes(entry.hash)) return entry
+    // Find the item's parents' indices
+    const indices = entry.next.map((next) => hashes.indexOf(next))
+    // Find the largest index (latest parent)
+    const index = indices.length > 0 ? Math.max(Math.max.apply(null, indices) + 1, 0) : 0
+    // Insert
+    log.items.splice(index, 0, entry)
+    return entry
+  }
+
+  /**
+   * Fetch log entries recursively
+   * @private
+   * @param {IPFS} [ipfs] An IPFS instance
+   * @param {string} [hash] Multihash of the entry to fetch
+   * @param {string} [parent] Parent of the node to be fetched
+   * @param {Object} [all] Entries to skip
+   * @param {Number} [amount=-1] How many entries to fetch.
+   * @param {Number} [depth=0] Current depth of the recursion
+   * @param {function(hash, entry, parent, depth)} onProgressCallback
+   * @returns {Promise<Array<Entry>>}
+   */
+  static _fetchRecursive(ipfs, hash, all = {}, amount = -1, depth = 0, parent = null, onProgressCallback = () => {}) {
+    // If the given hash is already fetched
+    // or if we're at maximum depth, return
+    if (all[hash] || (depth >= amount && amount > 0)) {
+      return Promise.resolve([])
+    }
+    // Create the entry and add it to the result
+    return Entry.fromMultihash(ipfs, hash)
+      .then((entry) => {
+        all[hash] = entry
+        onProgressCallback(hash, entry, parent, depth)
+        const fetch = (hash) => LogUtils._fetchRecursive(ipfs, hash, all, amount, depth + 1, entry, onProgressCallback)
+        return mapSeries(entry.next, fetch, { concurrency: 1 })
+          .then((res) => res.concat([entry]))
+          .then((res) => res.reduce((a, b) => a.concat(b), [])) // flatten the array
```

Comment:
it's probably more efficient to not generate a new array on every iteration, you can pull in `lodash.flatten` which is small and does it pretty efficiently.